FirebaseVertexAI/Sources/Types/Internal/ProtoDate.swift

```diff
@@ -0,0 +1,137 @@
+///
+/// This represents a
+/// [`google.type.Date`](https://cloud.google.com/vertex-ai/docs/reference/rest/Shared.Types/Date).
+struct ProtoDate {
+  /// Year of the date.
+  ///
+  /// Must be from 1 to 9999, or 0 to specify a date without a year.
+  let year: Int?
+
+  /// Month of a year.
+  ///
+  /// Must be from 1 to 12, or 0 to specify a year without a month and day.
+  let month: Int?
+
+  /// Day of a month.
+  ///
+  /// Must be from 1 to 31 and valid for the year and month, or 0 to specify a year by itself or a
+  /// year and month where the day isn't significant.
+  let day: Int?
+
+  /// Returns the a `DateComponents` representation of the `ProtoDate`.
+  ///
+  /// > Note: This uses the Gregorian `Calendar` to match the `google.type.Date` definition.
+  var dateComponents: DateComponents {
+    DateComponents(
+      calendar: Calendar(identifier: .gregorian),
+      year: year,
+      month: month,
+      day: day
+    )
+  }
+
+  /// Returns a `Date` representation of the `ProtoDate`.
+  ///
+  /// - Throws: An error of type `DateConversionError` if the `ProtoDate` cannot be represented as
+  ///   a `Date`.
+  func asDate() throws -> Date {
+    guard year != nil else {
+      throw DateConversionError(message: "Missing a year: \(self)")
+    }
+    guard month != nil else {
+      throw DateConversionError(message: "Missing a month: \(self)")
+    }
+    guard day != nil else {
+      throw DateConversionError(message: "Missing a day: \(self)")
+    }
+    guard dateComponents.isValidDate, let date = dateComponents.date else {
+      throw DateConversionError(message: "Invalid date: \(self)")
+    }
+    return date
+  }
+
+  struct DateConversionError: Error {
+    let localizedDescription: String
+
+    init(message: String) {
+      localizedDescription = message
+    }
+  }
+}
+
+// MARK: - Codable Conformance
+
+extension ProtoDate: Decodable {
+  enum CodingKeys: CodingKey {
+    case year
+    case month
+    case day
+  }
+
+  init(from decoder: any Decoder) throws {
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+    if let year = try container.decodeIfPresent(Int.self, forKey: .year), year != 0 {
+      guard year >= 1 && year <= 9999 else {
+        throw DecodingError.dataCorrupted(
+          .init(codingPath: [CodingKeys.year], debugDescription: "Invalid year: \(year)")
+        )
+      }
+      self.year = year
+    } else {
+      year = nil
+    }
+
+    if let month = try container.decodeIfPresent(Int.self, forKey: .month), month != 0 {
+      guard month >= 1 && month <= 12 else {
+        throw DecodingError.dataCorrupted(
+          .init(codingPath: [CodingKeys.month], debugDescription: "Invalid month: \(month)")
+        )
+      }
+      self.month = month
+    } else {
+      month = nil
+    }
+
+    if let day = try container.decodeIfPresent(Int.self, forKey: .day), day != 0 {
+      guard day >= 1 && day <= 31 else {
```

Comment:
What happens on an invalid month/day combination? Like 2024-09-31

Reply:
Hopefully the backend will never return such a combination but the [`dateComponents.isValidDate`](https://github.com/firebase/firebase-ios-sdk/pull/13893/commits/686f56b0ea7c98366ff3aa0120927cf8bf967e0f#diff-814a655daecaa001720e34d34870ad96bc64b7986c3db86bf4deee67f69d6295R64-R66) will check that and throw a `DateConversionError`. See [`testProtoDate_invalidDate_asDate_throws`](https://github.com/firebase/firebase-ios-sdk/pull/13893/commits/686f56b0ea7c98366ff3aa0120927cf8bf967e0f#diff-813f9d3277134fcf8afdb5fbc4b75547664bfa6a034ab706f64a13c908fdd349R136-R142) for an example (using November 31st). This error case will make it to the user when decoding the `Citation` [`431e794` (#13893)](https://github.com/firebase/firebase-ios-sdk/pull/13893/commits/431e794a8adc1558d411208ae33110b9d9fbdaf9#diff-c9f3458ead7c3fe7c0d7cfaf7caf241cf4cf54ae88496d017b764f0ac8dfdde8R400-R407).

Reply:
Note: Similar to above, this will now log a warning about an invalid date instead of throwing.